### PR TITLE
DOC-11959: Vector Search Code Snippets - iOS

### PR DESCRIPTION
**Pull request by velicuvlad**

### CODE REVIEW

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: These two lines shouldn't be part of the code snippet.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: I'm not sure if it will print out a useful info.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: If the scope of this code snippet is only for creating the default config, these 4 lines are not relevant.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: <vectorArray> will make the code snippet not being compiled. Also what is the scope of this code snippet? Does it need to show query as well?

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: This is incorrect. The model needs to implement CBLPredictiveModel protocol. Also, we will need to sync the code snippts for all platforms. I think we need some info or direction about the code snippet.

Anyway, here are two samples:

Samples:

// Color Model (I think server team has used RGB color as one of their sample apps)

```
@interface ColorModel : NSObject <CBLPredictiveModel>
@end

@implementation ColorModel

- (CBLDictionary*) predict: (CBLDictionary*)input {
   NSString* color = [input stringForKey: @"color"];
   NSArray* rgb = [RGBModel getRGBVectorForColorName: color]; // I made up the name of the model.
   
   CBLMutableDictionary* output = [[CBLMutableDictionary alloc] init];
   [output setValue: rgb forKey: @"vector"];
   return output;
}

@end
```

// Word Model 

```
@interface WordModel : NSObject <CBLPredictiveModel>
@end

@implementation WordModel

- (CBLDictionary*) predict: (CBLDictionary*)input {
   NSString* word = [input stringForKey: @"word"];
   NSArray* vector = [WordModel getVectorForWord: word]; // I made up the name of the model.
   
   CBLMutableDictionary* output = [[CBLMutableDictionary alloc] init];
   [output setValue: vector forKey: @"vector"];
   return output;
}

@end
```

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: swift is fine.
obj-c is giving the type and the mem address... `CBLScalarQuantizer:<...>`. I couldn't find a more suitable way to show the default values - I am open for suggestions

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: those are to show how to set custom settings

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: from my understanding is showing how to get the vector from Embedding and then how to pass it as params to the query

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: @ElliotFrancisHunter do we want/need to align with server on this?

Comment (ElliotFrancisHunter) [reply modules/objc/VectorSeachSnippets.m]: Aligning with server on examples is a good idea if we have customers using multiple products. I'm in favour of this choice

File: modules/objc/VectorSeachSnippets.m
Comment (ElliotFrancisHunter) [reply modules/objc/VectorSeachSnippets.m]: It is useful to show this, maybe specify with comments that these are custom settings, what they actually change for the end user etc

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: and what is server doing then?

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: there's line 45 to mark this. I dont think it needs extra code comments within the snippet... plain english documentation should be enough... 

all this information will be available within the api doc as well anyway, so I am not sure if is needed to show "how to assign value to property" at all @pasin

File: modules/objc/VectorSeachSnippets.m
Comment (ElliotFrancisHunter) [reply modules/objc/VectorSeachSnippets.m]: Server seems to have some demo code here https://docs.couchbase.com/server/current/vector-search/create-vector-search-index-rest-api.html#example

which is related to colour. For mobile let's use the word examples as it's simpler for integration into MVP of beta docs. We can refine them over time with more detail such as the colour model

**Review by ElliotFrancisHunter — APPROVED** [review]
I'm happy with these snippets. Plenty of useful stuff for me to work with 👍🏼 As long as @pasin is happy I'll merge and add these to the docs.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [reply modules/objc/VectorSeachSnippets.m]: PM suggested to me on the other day to have demo app converted to use the server's color model. I didn't have time to do it yet but the new sample app going forward will be color. It's ok to use word model as a sample now and convert it to use color later. It's not much of the change.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [reply modules/objc/VectorSeachSnippets.m]: @ElliotFrancisHunter I'll let you decide on this as it needs to be inline with the content structure.

File: modules/swift/VectorSearchSnippets.swift
Comment (pasin) [inline modules/swift/VectorSearchSnippets.swift]: The code from this line is not relavent. There is also `useVectorMatch()` that shows the query.

File: modules/swift/VectorSearchSnippets.swift
Comment (pasin) [inline modules/swift/VectorSearchSnippets.swift]: Again this is wrong. Please fix this for Swift as well.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: The code from this line is not relavent. There is already `useVectorMatch()` sample.

**Review by pasin — COMMENTED** [review]
I have just noticed this.

Should the code snippet is in the project here?

https://github.com/couchbase/docs-couchbase-lite/tree/release/3.1/modules/objc/examples
 
https://github.com/couchbase/docs-couchbase-lite/tree/release/3.1/modules/swift/examples
 
Also we have tried to make the code snippet compileable in those two XCode project. I think it's a good idea to follow that so that we don't make mistake in the code snippet.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: extra line

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: Move error to the line near `createIndexWithName`.

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: model needs to be created.

```
WordModel* model = [[WordModel alloc] init];
```

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [inline modules/objc/VectorSeachSnippets.m]: Change to use centorids = 8 for all for consistency.

**Review by pasin — CHANGES_REQUESTED** [review]

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [reply modules/objc/VectorSeachSnippets.m]: If we explain this in the doc content for this code snippet, we don't need to have all ofthe NSLog to show the default value at all.

Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: I do agree

**Comment by velicuvlad** [comment]
I know our current snippets structure, but my understanding is that Docs is looking to move away from the tags and was asked to have everything at the root of the platform directory. I have tested the snippets for interpreter and compile errors in terms of the api functionality and methods called.

To clarify, my expectations is that this won't be the final shape of the snippets, nor that the whole file to be added to the docs page quote on quote.

... will come back with another commit for some of the amends requested

**Comment by pasin** [comment]
If this is a new location, can you create an XCode project and add the file into the project? So that we don't need to manually copy the code and compile it somewhere else when we make change to the code.

**Comment by pasin** [comment]
You could aso add to the PR validation to verify the code snippet as well.

Here is an example in release/3.1 branch.
https://github.com/couchbase/docs-couchbase-lite/blob/release/3.1/.github/workflows/xcodebuild.yml

I'm asking this as in the past we made a lot of mistake in the code snippet.

You can have the structure like : 

```
root/<codesnippet.swift>
root/xcode/<Snippet.xcodeproject>
```

@ElliotFrancisHunter 

Can we use the existing project to keep the code snippet? I doesn't need to use the tags. The goal is to have the code within the compliable project to make sure that the code snippet can be compiled without an error. We aso have the Github PR validation to check that the code snippet is compiled. We do have the problems in the past that the code snippet had a lot of mistake and was not compilable.

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: I've address this by just using a declared placeholder

File: modules/objc/VectorSeachSnippets.m
Comment (velicuvlad) [reply modules/objc/VectorSeachSnippets.m]: will use word model for now

File: modules/objc/VectorSeachSnippets.m
Comment (ElliotFrancisHunter) [reply modules/objc/VectorSeachSnippets.m]: @pasin @velicuvlad the current position is good for the purposes of getting the content usable on time for the release. Hakim and I intend to refactor how the mobile docs codebase handles snippets so it's a bit more sustainable in future so there's no need for movement currently

File: modules/objc/VectorSeachSnippets.m
Comment (pasin) [reply modules/objc/VectorSeachSnippets.m]: I think it's better to describe it in the content instead of showing it as a comment in the code. So let remove it and add it back if we really need it.

**Comment by velicuvlad** [comment]
The validation automation must be done via Jenkins for multiple reasons such as access to the unreleased binaries etc. I have built the project on local and is fine.

This PR is ready for merge and can be merged if urgent until I find some time to take it with Jenkins - [CBL-5561](https://issues.couchbase.com/browse/CBL-5561) for tracking

**Comment by ElliotFrancisHunter** [comment]
> You could aso add to the PR validation to verify the code snippet as well.
> 
> Here is an example in release/3.1 branch. https://github.com/couchbase/docs-couchbase-lite/blob/release/3.1/.github/workflows/xcodebuild.yml
> 
> I'm asking this as in the past we made a lot of mistake in the code snippet.
> 
> You can have the structure like :
> 
> ```
> root/<codesnippet.swift>
> root/xcode/<Snippet.xcodeproject>
> ```
> 
> @ElliotFrancisHunter
> 
> Can we use the existing project to keep the code snippet? I doesn't need to use the tags. The goal is to have the code within the compliable project to make sure that the code snippet can be compiled without an error. We aso have the Github PR validation to check that the code snippet is compiled. We do have the problems in the past that the code snippet had a lot of mistake and was not compilable.

Yes that's fine. It'll make it easier for me to implement the snippets properly rather than the rough copy paste job I've done for sake of time.

**Comment by ElliotFrancisHunter** [comment]
@pasin good to merge for sake of release 👍🏼 just waiting on your review

File: modules/swift/examples/code_snippets/VectorSearch.swift
Comment (pasin) [inline modules/swift/examples/code_snippets/VectorSearch.swift]: extra line

File: modules/swift/examples/code_snippets/VectorSearch.swift
Comment (pasin) [inline modules/swift/examples/code_snippets/VectorSearch.swift]: extra line

File: modules/objc/examples/code_snippets/VectorSeach.m
Comment (pasin) [inline modules/objc/examples/code_snippets/VectorSeach.m]: extra line

File: modules/objc/examples/code_snippets/VectorSeach.m
Comment (pasin) [inline modules/objc/examples/code_snippets/VectorSeach.m]: extra line

File: modules/objc/examples/code_snippets/VectorSeach.m
Comment (pasin) [inline modules/objc/examples/code_snippets/VectorSeach.m]: extra line

File: modules/objc/examples/code_snippets/VectorSeach.m
Comment (pasin) [inline modules/objc/examples/code_snippets/VectorSeach.m]: extra line

File: modules/swift/examples/code_snippets/VectorSearch.swift
Comment (pasin) [inline modules/swift/examples/code_snippets/VectorSearch.swift]: extra lines

**Review by pasin — APPROVED** [review]
@velicuvlad I have some minor comments about the extra lines. The rest looks good to me.

**Comment by velicuvlad** [comment]
@ElliotFrancisHunter I'll merge this now

**Pull request merged**
